oap-server/server-alarm-plugin/src/test/java/org/apache/skywalking/oap/server/core/alarm/provider/NotifyHandlerTest.java

```diff
@@ -0,0 +1,202 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.oap.server.core.alarm.provider;
+
+import com.google.common.collect.Lists;
+import org.apache.skywalking.oap.server.core.CoreModule;
+import org.apache.skywalking.oap.server.core.alarm.AlarmMessage;
+import org.apache.skywalking.oap.server.core.alarm.MetaInAlarm;
+import org.apache.skywalking.oap.server.core.analysis.indicator.Indicator;
+import org.apache.skywalking.oap.server.core.analysis.indicator.IndicatorMetaInfo;
+import org.apache.skywalking.oap.server.core.analysis.indicator.WithMetadata;
+import org.apache.skywalking.oap.server.core.cache.EndpointInventoryCache;
+import org.apache.skywalking.oap.server.core.cache.ServiceInstanceInventoryCache;
+import org.apache.skywalking.oap.server.core.cache.ServiceInventoryCache;
+import org.apache.skywalking.oap.server.core.register.EndpointInventory;
+import org.apache.skywalking.oap.server.core.register.ServiceInstanceInventory;
+import org.apache.skywalking.oap.server.core.register.ServiceInventory;
+import org.apache.skywalking.oap.server.core.source.DefaultScopeDefine;
+import org.apache.skywalking.oap.server.library.module.ModuleManager;
+import org.apache.skywalking.oap.server.library.module.ModuleProviderHolder;
+import org.apache.skywalking.oap.server.library.module.ModuleServiceHolder;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.powermock.api.mockito.PowerMockito;
+import org.powermock.core.classloader.annotations.PowerMockIgnore;
+import org.powermock.core.classloader.annotations.PrepareForTest;
+import org.powermock.modules.junit4.PowerMockRunner;
+
+import static junit.framework.TestCase.assertNotNull;
+import static org.mockito.Mockito.*;
+
+/**
+ * Created by dengming, 2019.04.22
+ */
+@RunWith(PowerMockRunner.class)
+@PowerMockIgnore("javax.management.*")
+@PrepareForTest(DefaultScopeDefine.class)
+public class NotifyHandlerTest {
```

Comment:
In these tests, I don't find the check about `AlarmCallback` invoked. Do I miss something?

Reply:
What is the purpose of these tests? I am not very sure of that.

Reply:
1. In the NotifyHandler class( init() method), it just collects the AlarmCallback and then puts them into AlarmCore. So I think the AlarmCallback will be invoked in the AlarmCore class. In the NotifyHandlerTest, I just need to test that the NotifyHandler can collect the AlarmCallback and then put it into AlarmCore;
2.  Although some classes were tested by another Class's unit test indirectly, I think every Class which contains some logic should have the unit test that just tests the Class itself. That's the unit tests. So I add the unit tests for NotifyHandler, Threshold, AlarmModuleProvider. But I just find out that there is a unit test for RulesReader, so the RulesReaderTest will be removed.

Reply:
In my mind, if test `NotifyHandler#notify`, then we should check `MetaInAlarm` in different cases, because that is the core logic in this method. I found you mock different indicator, but doesn't check the result

> runningRules.forEach(rule -> rule.in(metaInAlarm, indicator));

Reply:
Yes, I don't check the ```rule.in(metaInAlarm, indicator)```. I think inside the NotifyHandlerTest, I just need to check the different branch due to different ```scope```.  I just simply check the code ```runningRules.forEach(rule -> rule.in(metaInAlarm, indicator));```  but not check the logic about ```MetaInAlarm```. 
The core logic you care about is tested in RunningRuleTest, is that right? RunningRuleTest checks the ```MetaInAlarm``` in different cases.
```java
// RunningRuleTest#testAlarm
runningRule.in(getMetaInAlarm(123), getIndicator(timeInPeriod1, 70));
runningRule.in(getMetaInAlarm(123), getIndicator(timeInPeriod2, 71));
runningRule.in(getMetaInAlarm(123), getIndicator(timeInPeriod3, 74));
```
Does RunningRuleTest miss something?

Reply:
I mean the metaInAlram argument for runningRule, when notify method invoked. That should be tested here. Right?

Reply:
en... Do you mean something like that:
```
doCheckAbout(metaInAlarm);
runningRules.forEach(rule -> rule.in(metaInAlarm, indicator));
```

Reply:
Yes

Reply:
Oh... I checked the ```metaInAlarm``` now.